Changes in module 'nf-core/merquryfk/merquryfk'
--- modules/nf-core/merquryfk/merquryfk/main.nf
+++ modules/nf-core/merquryfk/merquryfk/main.nf
@@ -30,6 +30,8 @@
     tuple val(meta), path("${prefix}.spectra-asm.ln.pdf") , emit: spectra_asm_ln_pdf, optional: true
     tuple val(meta), path("${prefix}.spectra-asm.st.png") , emit: spectra_asm_st_png, optional: true
     tuple val(meta), path("${prefix}.spectra-asm.st.pdf") , emit: spectra_asm_st_pdf, optional: true
+    tuple val(meta), path("${prefix}.false_duplications.tsv"), emit: false_duplications
+    tuple val(meta), path("${prefix}.cni.gz")                , emit: cn_histogram
     path "versions.yml"                                   , emit: versions
 
     when:
@@ -49,6 +51,12 @@
         $haplotigs \\
         $prefix
 
+    mv .cni ${prefix}.cni
+    awk -v asm_ploidy=${assembly instanceof List ? assembly.size() : 1} \\
+        -f $projectDir/bin/false_duplications.awk ${prefix}.cni \\
+        > ${prefix}.false_duplications.tsv
+    gzip ${prefix}.cni
+
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         fastk: $FASTK_VERSION

************************************************************
